### ISSUE TITLE
LTO issue on CRAN packages

### ISSUE THREAD

**Issue by boennecd** [r-lib/testthat#1230]
I recently submitted my psqn package on CRAN on 2020-11-04. However, I get an error with the LTO check because the `extern "C" SEXP run_testthat_tests()` has changed to: https://github.com/r-lib/testthat/blob/b684f1ee7fd806ff55a9d901174c4cc77c38b7d0/inst/include/testthat/testthat.h#L165

Thus, I get the following error on the LTO check on CRAN:

```bash
RcppExports.cpp:54:17: warning: ‘run_testthat_tests’ violates the C++ One Definition Rule [-Wodr]
   54 | RcppExport SEXP run_testthat_tests();
      |                 ^
/data/gannet/ripley/R/test-4.1/testthat/include/testthat/testthat.h:165:17: note: type mismatch in parameter 1
  165 | extern "C" SEXP run_testthat_tests(SEXP use_xml_sxp) {
      |                 ^
```

after the testthat release at the end of October. This must break the LTO check on all packages that use catch through testthat. I cannot see that this is addressed anywhere (I am sorry if I have missed something!). Moverer, 
 
 1. installing the most recent version of Rcpp and testthat,
 2. deleting `RcppExports.cpp`, and
 3. running `Rcpp::compileAttributes()`

still incorrectly yields:

```c++
RcppExport SEXP run_testthat_tests();
``` 

in `RcppExports.cpp`. I guess a solution in the short term is to overwrite the line with:

```c++
RcppExport SEXP run_testthat_tests(SEXP use_xml_sxp);

// ...

static const R_CallMethodDef CallEntries[] = {
    // ...
    {"run_testthat_tests", (DL_FUNC) &run_testthat_tests, 1 /* was zero before */},
    // ...
};
```

and change `tests/testthat/test-cpp.R` to:

```r
(function() {
  .Call("run_testthat_tests", FALSE, PACKAGE = "psqn")
})
```

This seems to work for me.

**Comment by gaborcsardi** [comment]
Yeah, it seems that this is from https://github.com/r-lib/testthat/pull/1008

It breaks the LTO checks for ~ 11 CRAN packages, only the ones that use `TESTTHAT_TEST_RUNNER` seem to be affected: https://github.com/search?p=1&q=org%3Acran+TESTTHAT_TEST_RUNNER&type=Code

**Comment by hadley** [comment]
@jimhester can you take a look please?

**Comment by hadley** [comment]
I think everyone that this affects has been notified so we can close the issue.

**Comment by bitbacchus** [comment]
We are in the process of publishing our package to CRAN and stumbled over this issue. Just to clarify that I understood this correctly: I have to manually edit `RcppExports.cpp` and `tests/testthat/test-cpp.R` to circumvent the issue? 

Or is there a fix? I have Rcpp 1.0.6 and testthat 3.0.3 installed and after deleting and recreating my `RcppExports.cpp` I still have this LTO problem.

Thanks,

Sebastian

**Comment by slwu89** [comment]
@jimhester is there any update on this? I know you mentioned manually changing RcppExports.cpp but as that's an autogenerated file that method to deal with this issue seems very brittle. Thanks!
